File: 2_ArmacodeBeta2/java/PlantTrees.java
import java.util.*;

public class OxygenLevels {

    static class TupleL3 implements Comparable<TupleL3> {
        long first, second, third;

        TupleL3(long a, long b, long c) {
            first = a;
            second = b;
            third = c;
        }

        @Override
        public int compareTo(TupleL3 other) {
            return Long.compare(this.second, other.second);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int N = scanner.nextInt();
        TupleL3[] vals = new TupleL3[N];

        for (int i = 0; i < N; ++i) {
            long i_start = scanner.nextLong();
            long i_end = scanner.nextLong();
            long i_oxygen = scanner.nextLong();
            vals[i] = new TupleL3(i_start, i_end, i_oxygen);
        }

        Arrays.sort(vals);

        TreeMap<Long, Long> oxygenSoFar = new TreeMap<>();
        oxygenSoFar.put(0L, 0L);
        long maxOxygen = 0;

        for (int i = 0; i < N; ++i) {
            long start = vals[i].first;
            long end = vals[i].second;
            long oxy = vals[i].third;

            Map.Entry<Long, Long> lowerBound = oxygenSoFar.floorEntry(start - 1);

            oxy += lowerBound.getValue();
            maxOxygen = Math.max(oxy, maxOxygen);
            oxygenSoFar.put(end, maxOxygen);
        }

        System.out.println(maxOxygen);
        scanner.close();
    }
}

File: 2_ArmacodeBeta2/cpp/PlantTrees.cpp
#include <cmath>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
using namespace std;

typedef long long intLL;
typedef tuple<intLL, intLL, intLL> tuple_ll3;

bool compVals(const tuple_ll3 & a, const tuple_ll3 & b) {
    return get<1>(a) < get<1>(b);
}

int main() {
    long N;
    cin >> N;
    vector<tuple_ll3> vals(N);
    intLL i_start, i_end, i_oxygen;
    for (long i = 0; i < N; ++i) {
        cin >> i_start >> i_end >> i_oxygen;
        vals[i] = {i_start, i_end, i_oxygen};
    }
    sort(
        vals.begin(), vals.end(), compVals
    );

    map<intLL, intLL> oxygenSoFar;
    oxygenSoFar[0] = 0;
    intLL maxOxygen = 0;

    for(long i = 0; i < N; ++i) {
        intLL start = get<0>(vals[i]);
        intLL end   = get<1>(vals[i]);
        intLL oxy   = get<2>(vals[i]);
        auto lowerBound = -- oxygenSoFar.lower_bound(start);

        oxy += lowerBound->second;
        maxOxygen = max(oxy, maxOxygen);
        oxygenSoFar[end] = maxOxygen;
    }

    cout << maxOxygen;
    return 0;
}